internal/peering-request-operator/foreign-cluster.go

```diff
@@ -0,0 +1,158 @@
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+	"k8s.io/utils/pointer"
+	"strings"
+)
+
+func (r *PeeringRequestReconciler) UpdateForeignCluster(pr *v1.PeeringRequest) error {
+	tmp, err := r.crdClient.Resource("foreignclusters").Get(pr.Spec.ClusterID, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		// try to find it by label
+		err2 := err
+		tmp, err = r.crdClient.Resource("foreignclusters").List(metav1.ListOptions{
+			LabelSelector: "cluster-id=" + pr.Spec.ClusterID,
+		})
+		if err != nil {
+			klog.Error(err, err.Error())
+			return err
+		}
+		fcList, ok := tmp.(*v1.ForeignClusterList)
+		if !ok {
+			err = errors.New("retrieved object is not a ForeignClusterList")
+			klog.Error(err, err.Error())
+			return err
+		}
+		if len(fcList.Items) == 0 {
+			// object not found
+			err = err2
+		} else {
+			// objects retrieved
+			err = nil
+			tmp = &fcList.Items[0]
+		}
+	}
+
+	if k8serrors.IsNotFound(err) {
+		// create it
+		fc, err := r.createForeignCluster(pr)
+		if err != nil {
+			return err
+		}
+		r.setOwner(pr, fc)
+		return nil
+	} else if err != nil {
+		// other errors
+		klog.Error(err, err.Error())
+		return err
+	} else {
+		// update it
+		fc, ok := tmp.(*v1.ForeignCluster)
+		if !ok {
+			err = errors.New("retrieved object is not a ForeignCluster")
+			klog.Error(err, err.Error())
+			return err
+		}
+		if fc.Status.PeeringRequest != nil {
+			// already up to date
+			return nil
+		}
+		fc.Status.PeeringRequest = &corev1.ObjectReference{
+			Kind:       pr.Kind,
+			Name:       pr.Name,
+			UID:        pr.UID,
+			APIVersion: pr.APIVersion,
+		}
+		_, err = r.crdClient.Resource("foreignclusters").Update(fc.Name, fc, metav1.UpdateOptions{})
+		if err != nil {
+			klog.Error(err, err.Error())
+			return err
+		}
+		r.setOwner(pr, fc)
+		return nil
+	}
+}
+
+func (r *PeeringRequestReconciler) createForeignCluster(pr *v1.PeeringRequest) (*v1.ForeignCluster, error) {
+	var cnf *rest.Config
+	var err error
+	if r.ForeignConfig != nil {
+		// testing
+		cnf = r.ForeignConfig
+	} else {
+		cnf, err = pr.GetConfig(r.crdClient.Client())
+		if err != nil {
+			klog.Error(err, err.Error())
+			return nil, err
+		}
+	}
+
+	fc := &v1.ForeignCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: pr.Spec.ClusterID,
+		},
+		Spec: v1.ForeignClusterSpec{
+			ClusterID:        pr.Spec.ClusterID,
+			Namespace:        pr.Spec.Namespace,
+			Join:             r.getReverseJoin(pr),
+			ApiUrl:           cnf.Host,
+			DiscoveryType:    v1.IncomingPeeringDiscovery,
+			AllowUntrustedCA: pr.Spec.OriginClusterSets.AllowUntrustedCA,
+		},
+		Status: v1.ForeignClusterStatus{
+			PeeringRequest: &corev1.ObjectReference{
+				Kind:       pr.Kind,
+				Name:       pr.Name,
+				UID:        pr.UID,
+				APIVersion: pr.APIVersion,
+			},
+		},
+	}
+
+	if r.ForeignConfig != nil {
+		fc.Spec.ApiUrl = strings.Replace(fc.Spec.ApiUrl, "https://", "http://", -1)
```

Comment:
Why do we need this replace? Can't we assume that the ApiURL is https?

Reply:
I'm sorry, I forgot to add a comment. r.ForeignConfig is only set in test environment (where we have only one ForeignConfig) and in this environment API server is only available in http, but field is automatically created with https, because yes, we can assume that ApiUrl is https

Reply:
We should avoid at most to add code for testing purposes, and put it in fake classes.